Add retry config for failing steps (#2627)
zenml-io/zenml: docs/book/user-guide/advanced-guide/pipelining-features/managing-steps.md

Reviewer: My first question here as I read it was 'does ZenML support / allow infinite retries', so maybe you can answer that in the documentation?

2. Disable it by using the environmental variable `ZENML_DISABLE_STEP_LOGS_STORAGE` and setting it to `true`. This environmental variable takes precedence over the parameters mentioned above.

<figure><img src="https://static.scarf.sh/a.png?x-pxid=f0b4f458-0a54-4fcd-aa95-d5ee424815bc" alt="ZenML Scarf"><figcaption></figcaption></figure>

## Allow step retry in case of failure

> ZenML provides a built-in retry mechanism that allows you to configure automatic retries for your steps in case of failures. This can be useful when dealing with intermittent issues or transient errors.

Reviewer: ```suggestion
ZenML provides a built-in retry mechanism that allows you to configure automatic retries for your steps in case of failures. This can be useful when dealing with intermittent issues or transient errors. A common pattern when trying to run a step on GPU-backed hardware is that the provider will not have enough resources available, so you can set ZenML to handle the retries until the resources free up.
```

> You can configure the retry behavior for a step using the retry parameter in the @step decorator or through a configuration file:

Reviewer: ```suggestion
You can configure the retry behavior for a step using the retry parameter in the `@step` decorator or through a configuration file:
```

### Using the @step decorator:

```python
from zenml.config.retry_config import StepRetryConfig

@step(
    retry=StepRetryConfig(
        max_retries=3, 
        delay=10, 
        backoff=2
    )
)
def my_step() -> None:
    raise Exception("This is a test exception")
```

* **max_retries:** The maximum number of times the step should be retried in case of failure.
* **delay:** The initial delay in seconds before the first retry attempt.
* **backoff:** The factor by which the delay should be multiplied after each retry attempt.

### Using a configuration file (e.g., config.yaml):

```yaml
steps:
  my_step:
    retry:
      max_retries: 3
      delay: 10
      backoff: 2
```

> 

Reviewer: Maybe add a line here that links back to the 'how to use config.yaml' section / page of the docs. It also in general needs just a bit of text alongside the code.

2. Disable it by using the environmental variable `ZENML_DISABLE_STEP_LOGS_STORAGE` and setting it to `true`. This environmental variable takes precedence over the parameters mentioned above.

<figure><img src="https://static.scarf.sh/a.png?x-pxid=f0b4f458-0a54-4fcd-aa95-d5ee424815bc" alt="ZenML Scarf"><figcaption></figcaption></figure>

## Allow step retry in case of failure

ZenML provides a built-in retry mechanism that allows you to configure automatic retries for your steps in case of failures. This can be useful when dealing with intermittent issues or transient errors. A common pattern when trying to run a step on GPU-backed hardware is that the provider will not have enough resources available, so you can set ZenML to handle the retries until the resources free up. You can configure three parameters for step retries:

* **max_retries:** The maximum number of times the step should be retried in case of failure.
* **delay:** The initial delay in **seconds** before the first retry attempt.
* **backoff:** The factor by which the delay should be multiplied after each retry attempt.

> ### Using the @step decorator:
> 
> ```python
> from zenml.config.retry_config import StepRetryConfig
> 
> @step(
>     retry=StepRetryConfig(
>         max_retries=3, 
>         delay=10, 
>         backoff=2
>     )
> )
> def my_step() -> None:
>     raise Exception("This is a test exception")
> ```
> 
> You can also configure step retry using the [Configuration YAML](https://docs.zenml.io/user-guide/production-guide/configure-pipeline#breaking-down-our-configuration-yaml).
> Just like the Pythonic version we set the retry config in a `steps` section where we specify the name of the step and the retry configuration.
> ### Using a configuration file (e.g., config.yaml):
> 
> ```yaml

Reviewer: ```suggestion
### Using the @step decorator:

You can specify the retry configuration directly in the definition of your step as follows:

```python
from zenml.config.retry_config import StepRetryConfig

@step(
    retry=StepRetryConfig(
        max_retries=3, 
        delay=10, 
        backoff=2
    )
)
def my_step() -> None:
    raise Exception("This is a test exception")
```

### Using a configuration file (e.g., config.yaml):

You can also configure step retry using the [Configuration YAML](https://docs.zenml.io/user-guide/production-guide/configure-pipeline#breaking-down-our-configuration-yaml).
Just like the Pythonic version we set the retry config in a `steps` section where we specify the name of the step and the retry configuration.

```yaml
```

## Allow step retry in case of failure

> ZenML provides a built-in retry mechanism that allows you to configure automatic retries for your steps in case of failures. This can be useful when dealing with intermittent issues or transient errors. A common pattern when trying to run a step on GPU-backed hardware is that the provider will not have enough resources available, so you can set ZenML to handle the retries until the resources free up. You can configure 3 parameter for step retry:

Reviewer: ```suggestion
ZenML provides a built-in retry mechanism that allows you to configure automatic retries for your steps in case of failures. This can be useful when dealing with intermittent issues or transient errors. A common pattern when trying to run a step on GPU-backed hardware is that the provider will not have enough resources available, so you can set ZenML to handle the retries until the resources free up. You can configure three parameters for step retries:
```

    raise Exception("This is a test exception")
```

> You can also configure step retry using [Configuration YAML](https://docs.zenml.io/user-guide/production-guide/configure-pipeline#breaking-down-our-configuration-yaml)

Reviewer: ```suggestion
You can also configure step retry using the [Configuration YAML](https://docs.zenml.io/user-guide/production-guide/configure-pipeline#breaking-down-our-configuration-yaml).
```

> just like the pythonic version we set the retry config in a `steps` section where we specify the name of the step and the retry configuration.

Reviewer: ```suggestion
Just like the Pythonic version we set the retry config in a `steps` section where we specify the name of the step and the retry configuration.
```

```

{% hint style="info" %}

> Note that infinite retries are not supported at the moment. If you set max_retries to a very large value or do not specify it at all, ZenML will still enforce an internal maximum number of retries to prevent infinite loops. We recommend setting a reasonable max_retries value based on your use case and the expected frequency of transient failures.

Reviewer: ```suggestion
Note that infinite retries are not supported at the moment. If you set `max_retries` to a very large value or do not specify it at all, ZenML will still enforce an internal maximum number of retries to prevent infinite loops. We recommend setting a reasonable `max_retries` value based on your use case and the expected frequency of transient failures.
```

What is 'a very large value'?
